bundles/io/org.eclipse.smarthome.io.rest.sitemap/src/main/java/org/eclipse/smarthome/io/rest/sitemap/internal/SitemapResource.java

```diff
@@ -614,9 +616,55 @@ private boolean waitForChanges(EList<Widget> widgets) {
                     } catch (ItemNotFoundException e) {
                         // ignore
                     }
-                } else {
-                    if (widget instanceof Frame) {
-                        items.addAll(getAllItems(((Frame) widget).getChildren()));
+                }
+                // Consider all items inside the frame
+                if (widget instanceof Frame) {
+                    items.addAll(getAllItems(((Frame) widget).getChildren()));
+                }
+                // Consider items involved in any visibility, labelcolor and valuecolor condition
+                items.addAll(getItemsInVisibilityCond(widget.getVisibility()));
+                items.addAll(getItemsInColorCond(widget.getLabelColor()));
+                items.addAll(getItemsInColorCond(widget.getValueColor()));
+            }
+        }
+        return items;
+    }
+
+    private Set<GenericItem> getItemsInVisibilityCond(EList<VisibilityRule> ruleList) {
+        Set<GenericItem> items = new HashSet<GenericItem>();
+        if (ruleList != null) {
```

Comment:
`rulesList` will never be null, `widget#getVisibility` always returns a list instance.